README.md
# MMM-LoLeSports

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

This Module displays upcoming matches for eSports leagues. The Data is retrieved from [Pandascore](https://pandascore.co/).
The Module should support all Games that are delivered by Pandascore. I only tested League of Legends so far. No guarantee that it will work the other games as LoL. The free Pandascore-Account allowes up to 1000 requests per Hour.

![only text](https://github.com/MartinGaiser/MMM-eSports/blob/master/.github/onlyText.jpg "Only Text")
![only images](https://github.com/MartinGaiser/MMM-eSports/blob/master/.github/onlyImages.jpg "Only Images")
![header](https://github.com/MartinGaiser/MMM-eSports/blob/master/.github/onlyImages.jpg "Header")

## Installation
1. Clone this repo into the ```~/MagicMirror/modules```. 
2. Go into the Module direcroty: ```cd MMM-eSports```
3. Run ```npm install```
4. Configure your ```~/MagicMirror/config/config.js```:
    ```js
       {
            module: 'MMM-eSports',
            position: 'top_right',
            header: 'Custom Header', //Optional
            config: {
                ...
            }
        }
     ```

## Configuration options

| Option           |Default| Description
|----------------- |-------|-----------
| `apiKey`         |   -   | **Required** To get the API-Key create a free [Pandascore Account](https://pandascore.co/users/sign_up). 
| `leagueIDs`     |   -   | **Required** You will need to find the leagueIDs of the Leagues you want to display first. To do that use the [GetLeagues](https://developers.pandascore.co/doc/#operation/get_leagues)-Endpoint. Don't forget to supply your Token! 
| `updateInterval` | 60    | Time between requests in seconds. 
| `numberOfGames`  | 10    | The number of Matches that will be displayed (number of rows)
| `language`       | 'en'  | The language, all languages of moment.js are supported
| `timeFormat`     | 12    | Either 12 (pm/am) or 24
| `leagueAsImage`  | false | If true the league icon will be displayed instead of the leagues name
| `teamAsImage`    | false | If true the teams icon will be displayed instead of the teams name
| `showHeader` | false | Show the custom Header (Calendar Icon, Team1, Team2 League. See Image 'Custom Header')


Example Config: 

```js
config: {
    updateInterval: 60,         
    apiKey: "<YOUR-API-KEY>",   
    leagueIDs: "4197,4302",    
    numberOfGames: 5,           
    language: "en",             
    timeFormat: 12,            
    leagueAsImage: true,        
    teamAsImage: true,          
    size: "80%"                
}
```

### Finding League IDs
Example for finding the ID of League of Legends LEC:
```
https://api.pandascore.co/lol/leagues?search[name]=LEC&token=<yourToken>
```
Rsponse: 
```
[
    {
        "id": 4197,
        "image_url": "https://cdn.pandascore.co/images/league/image/4197/LEC-2019.png",
        "modified_at": "2019-03-03T01:22:54Z",
        "name": "LEC",
        .
        .
        .
    }
]
```
The ID of LEC therefore is: 4197.


CHANGELOG.md
# MMM-eSports Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Unreleased

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-eSports
 *
 * By Martin Gaiser
 * MIT Licensed.
 */
var request = require("request");
var NodeHelper = require("node_helper");
var moment = require("moment");

module.exports = NodeHelper.create({

	// Override socketNotificationReceived method.

	/* socketNotificationReceived(notification, payload)
	 * This method is called when a socket notification arrives.
	 *
	 * argument notification string - The identifier of the noitication.
	 * argument payload mixed - The payload of the notification.
	 */
	socketNotificationReceived: function(notification, payload) {
		if (notification === "MMM-eSports-StartFetching") {
			console.log("Starting to Fetch League Matches");
			let config = payload;
			moment.updateLocale(config.language, this.getLocaleSpecification(config.timeFormat));
			//Start Interval Fething of Data
			this.getData(config.apiKey, config.numberOfGames, config.leagueIDs, config.updateInterval*1000, 2000);
		}
	},

	getData: function(apiKey, xPerPage, leagueIDs, updateDelay, backoff) {
		console.log("Fetching new Data from Pandascore");
		let urlApi = "https://api.pandascore.co/lol/matches/upcoming";
		urlApi = urlApi.concat("?filter[league_id]=" + leagueIDs);
		urlApi = urlApi.concat("&sort=scheduled_at");
		urlApi = urlApi.concat("&page=1");
		urlApi = urlApi.concat("&per_page=" + xPerPage);
		const options = {
			url: urlApi,
			headers: {
			  "Authorization": "Bearer " + apiKey,
			}
		};

		var self = this;


		const callback = function(error, response, body){
			if (!error && response.statusCode == 200) {
				self.sendLeagueDataNotification(JSON.parse(body)); //send Data to Frontend
				self.retryTimeout = self.defaultRetryTimeout; //reset retry timeout after successful request
				setTimeout(function(){	//start Timeout for new request
					self.getData(apiKey, xPerPage, leagueIDs, updateDelay);
				}, updateDelay);
			}else if (!error && response.statusCode == 401){
				self.sendUnauthorizedNotification(); // send unauthorized message to frontned and stop loop
			}else{
				console.log("MMM-eSports: StatusCode == " + response.statusCode);
				self.sendErrorNotification(backoff); //send error message to frontend
				setTimeout(function(){ //set retry Timeout
					self.getData(apiKey, xPerPage, leagueIDs, updateDelay, backoff*2);
				}, backoff);
			}
		}
		request(options, callback);
	},

	// Example function send notification test
	sendUnauthorizedNotification: function(payload) {
		this.sendSocketNotification("MMM-eSports-Unauthorized", payload);
	},

	sendErrorNotification: function(payload) {
		this.sendSocketNotification("MMM-eSports-Error",payload);
	},

	sendLeagueDataNotification: function(payload) {
		let numberOfMatches = Object.keys(payload).length;
		let matches = [];
		let match = null;
		for (i = 0; i < numberOfMatches-1; i++){
			let leagueName = payload[i].league.name;
			let scheduledAt = payload[i].scheduled_at;
			let team1 = payload[i].opponents[0].opponent.acronym;
			let team2 = payload[i].opponents[1].opponent.acronym;
			let team1Url = payload[i].opponents[0].opponent.image_url;
			let team2Url = payload[i].opponents[1].opponent.image_url;
			let leagueImage = payload[i].league.image_url;
			let myMoment = moment(scheduledAt);
			if (moment(scheduledAt).diff(moment(), 'days', true) <= 1){
				scheduledAt = myMoment.fromNow();
			}else{
				scheduledAt = myMoment.calendar();
			}
			
			match = {
				"leagueName": leagueName,
				"scheduledAt": scheduledAt,
				"team1": team1,
				"team2": team2,
				"leagueImage": leagueImage,
				"team1Url": team1Url,
				"team2Url": team2Url
			}
			matches.push(match);
		}
		this.sendSocketNotification("MMM-eSports-GameData", matches);
	},

	getLocaleSpecification: function(timeFormat) {
		switch (timeFormat) {
		case 12: {
			return { longDateFormat: {LT: "h:mm A"} };
		}
		case 24: {
			return { longDateFormat: {LT: "HH:mm"} };
		}
		default: {
			return { longDateFormat: {LT: moment.localeData().longDateFormat("LT")} };
		}
		}
	},

});


MMM-LoLeSports.js
/* global Module */

/* Magic Mirror
 * Module: MMM-LoLeSports
 *
 * By Martin Gaiser
 * MIT Licensed.
 */

Module.register("MMM-LoLeSports", {
	defaults: {
		updateInterval: 350000,
		size: "60%",
		apiKey: "nokey",
		league_ids: "4302",
		numberOfGames: 10,
		timeFormat: 12,
		language: "en",
		leagueAsImage: false,
		teamAsImage: false
	},

	requiresVersion: "2.1.0", // Required version of MagicMirror

	start: function() {
		Log.info("Starting module: " + this.name);
		var self = this;
		var leagueData = null;
		var error = false;
		var unauthorized = false;
	},

	getDom: function() {
		var self = this;

		if (this.error){
			var wrapper = document.createElement("div");
			wrapper.innerHTML = "Unknown Error ... check logs.";
			return wrapper;
		}

		if (this.unauthorized){
			var wrapper = document.createElement("div");
			wrapper.innerHTML = "Unauthorized ...check API-Token";
			return wrapper;
		}

		// If this.dataRequest is not empty
		if (this.leagueData == null) {
			var wrapper = document.createElement("div");
			wrapper.innerHTML = "Loading...";
			this.sendSocketNotification("MMM-LoLeSports-StartFetching", this.config)
			return wrapper;
		}else{
			var wrapper = document.createElement("table");
			wrapper.classList.add("leaguetable")
			wrapper.setAttribute("font-size", this.config.size)
			wrapper.appendChild(this.getHeaderRow());
			for (let i = 0; i < this.leagueData.length; i++){
				wrapper.appendChild(this.getDataRow(this.leagueData[i].scheduled_at, this.leagueData[i].leagueName, this.leagueData[i].leagueImage,
					 this.leagueData[i].team1,this.leagueData[i].team1Url, this.leagueData[i].team2, this.leagueData[i].team2Url))
			}
			return wrapper;
		}
	},

	socketNotificationReceived: function(notification, payload){
		if (notification == "MMM-LoLeSports-Unauthorized"){
			this.unauthorized = true;
			this.updateDom();
		}
		if (notification == "MMM-LoLeSports-Error"){
			this.error = true;
			this.updateDom();
		}
		if (notification == "MMM-LoLeSports-GameData"){
			this.leagueData = payload;
			this.updateDom();
		}
	},

	getScripts: function () {
		return ["moment.js"];
	},

	getStyles: function () {
		return [
			"MMM-LoLeSports.css",
			"font-awesome.css",
		];
	},

	// Load translations files
	getTranslations: function() {
		//FIXME: This can be load a one file javascript definition
		return {
			en: "translations/en.json",
			es: "translations/es.json"
		};
	},

	getHeaderRow: function(){
		let header = document.createElement("tr");

		let dataCell = document.createElement("th");
		dataCell.classList.add("dateheader", "th");
		
		dataCell.classList.add("leagueheader", "th");
		let dateIcon = document.createElement('i');
        dateIcon.classList.add('fa', 'fa-calendar');
		dataCell.appendChild(dateIcon);
		header.appendChild(dataCell);

		dataCell = document.createElement("th");
		dataCell.classList.add("leagueheader", "th");
		header.appendChild(dataCell);

		dataCell = document.createElement("th");
		dataCell.classList.add("team1header", "th");
		dataCell.innerHTML = "Team 1";
		header.appendChild(dataCell);

		dataCell = document.createElement("th");
		dataCell.classList.add("versusheader","th");
		header.appendChild(dataCell);

		dataCell = document.createElement("th");
		dataCell.classList.add("team2header","th");
		dataCell.innerHTML = "Team 2 ";
		header .appendChild(dataCell);
		return header;
	},

	getVersusCell: function(){
		let versusCell = document.createElement("td");
		versusCell.classList.add("versuscell", "td");
		versusCell.innerHTML = "VS"
		return versusCell;
	},

	getDataRow: function(date, league, leagueImageURL, team1,team1Url, team2, team2Url){
		let row = document.createElement("tr");

		
		let dataCell = document.createElement("td")
		dataCell.classList.add("datecell","td");
		dataCell.innerHTML = date;
		row.appendChild(dataCell);

		if(this.config.leagueAsImage){
			row.appendChild(this.getLeagueAsImage(leagueImageURL));
		}else{
			row.appendChild(this.getLeagueAsText(league));
		}

		if(this.config.teamAsImage){
			row.appendChild(this.getTeamAsImage(team1Url));
		}else{
			row.appendChild(this.getTeamAsText(team1));
		}
		row.appendChild(this.getVersusCell());
		if(this.config.teamAsImage){
			row.appendChild(this.getTeamAsImage(team2Url));
		}else{
			row.appendChild(this.getTeamAsText(team2));
		}

		return row;
	},

	getLeagueAsImage: function(iconUrl){
		let dataCell = document.createElement("td");
		dataCell.classList.add("leagueCell", "td");
		let image = document.createElement("img");
		image.classList.add("leagueIcon", "img");
		image.setAttribute("src", iconUrl);
		dataCell.appendChild(image);
		return dataCell;
	},

	getLeagueAsText: function(leagueText){
		let dataCell = document.createElement("td");
		dataCell.classList.add("leagueCell","td");
		dataCell.innerHTML = leagueText;
		return dataCell;

	},

	getTeamAsImage: function(iconUrl){
		let dataCell = document.createElement("td");
		dataCell.classList.add("teamIconCell", "td");
		let image = document.createElement("img");
		image.classList.add("teamIcon", "img");
		image.setAttribute("src", iconUrl);
		dataCell.appendChild(image);
		return dataCell;
	},

	getTeamAsText: function(teamText){
		dataCell = document.createElement("td");
		dataCell.innerHTML = teamText;
		return dataCell;
	}
});


MMM-eSports.js
/* global Module */

/* Magic Mirror
 * Module: MMM-eSports
 *
 * By Martin Gaiser
 * MIT Licensed.
 */

Module.register("MMM-eSports", {
	defaults: {
		updateInterval: 350000,
		apiKey: "nokey",
		leagueIDs: "4302",
		numberOfGames: 10,
		timeFormat: 12,
		language: "en",
		leagueAsImage: false,
		teamAsImage: false,
		showHeader: false,
		showTimestamp:false
	},

	requiresVersion: "2.1.0", // Required version of MagicMirror

	start: function() {
		Log.info("Starting module: " + this.name);
		this.leagueData = null;
		this.error = false;
		this.unauthorized = false;
		this.errorRetry = 0;
	},

	getDom: function() {
		if (this.error){
			var wrapper = document.createElement("div");
			this.errorRetry = this.errorRetry-1;
			wrapper.innerHTML = this.translate("UNKNOWN-ERROR") + this.errorRetry + "s";
			return wrapper;
		}

		if (this.unauthorized){
			var wrapper = document.createElement("div");
			wrapper.innerHTML = this.translate("UNAUTHORIZED");
			return wrapper;
		}

		// If this.leagueData is empty, assume module just started. Tell Backend to start fetching Data
		if (this.leagueData == null) {
			var wrapper = document.createElement("div");
			wrapper.innerHTML = this.translate("LOADING");
			this.sendSocketNotification("MMM-eSports-StartFetching", this.config)
			return wrapper;
		//If this.leagueData is not empty, display the data
		}else{
			var wrapper = document.createElement("div");
			var table = document.createElement("table");
			wrapper.appendChild(table);
			table.classList.add("leaguetable");
			if (this.config.showHeader){
				table.appendChild(this.getHeaderRow());
			}
			for (let i = 0; i < this.leagueData.length; i++){
				table.appendChild(this.getDataRow(this.leagueData[i].scheduledAt, this.leagueData[i].leagueName, this.leagueData[i].leagueImage,
					 this.leagueData[i].team1,this.leagueData[i].team1Url, this.leagueData[i].team2, this.leagueData[i].team2Url))
			}
			if (this.config.showTimestamp){
				wrapper.appendChild(this.getTimeStamp());
			}
			return wrapper;
		}
	},

	socketNotificationReceived: function(notification, payload){
		if (notification == "MMM-eSports-Unauthorized"){
			clearInterval();
			this.unauthorized = true;
			this.updateDom();
		}
		if (notification == "MMM-eSports-Error"){
			var self = this;
			var IsInErrorState = this.error;
			this.error = true;
			this.errorRetry = payload/1000; // format to seconds
			if (!IsInErrorState){
				Log.log("Starting Interval");
				setInterval(function(){
					self.updateDom();
				}, 1000); //Update Fronent each second and to count down retry timer
			}
		}
		if (notification == "MMM-eSports-GameData"){
			clearInterval();
			this.leagueData = payload;
			this.unauthorized = false;
			this.error = false;
			this.errorRetry = 0;
			this.updateDom();	//Update Dom
		}
	},

	getScripts: function () {
		return ["moment.js"];
	},

	getStyles: function () {
		return [
			"MMM-eSports.css",
			"font-awesome.css",
		];
	},

	// Load translations files
	getTranslations: function() {
		//FIXME: This can be load a one file javascript definition
		return {
			en: "translations/en.json",
			es: "translations/es.json"
		};
	},

	getTimeStamp: function(){
		let table = document.createElement("table");
		let timestmap = document.createElement("div");
		timestmap.classList.add("small", "dimmed");
		timestmap.innerHTML = moment().calendar();
		table.appendChild(document.createElement("tr").appendChild(document.createElement("td").appendChild(timestmap)));
		return table;		
	},

	getHeaderRow: function(){
		let header = document.createElement("tr");

		let dataCell = document.createElement("th");
		dataCell.classList.add("dateheader", "th");

		dataCell.classList.add("leagueheader", "th");
		let dateIcon = document.createElement("i");
		dateIcon.classList.add("fa", "fa-calendar");
		dataCell.appendChild(dateIcon);
		header.appendChild(dataCell);

		dataCell = document.createElement("th");
		dataCell.classList.add("leagueheader", "th");
		dataCell.innerHTML = this.translate("LEAGUE");
		header.appendChild(dataCell);

		dataCell = document.createElement("th");
		dataCell.classList.add("team1header", "th");
		dataCell.innerHTML = "Team 1";
		header.appendChild(dataCell);

		dataCell = document.createElement("th");
		dataCell.classList.add("versusheader","th");
		header.appendChild(dataCell);

		dataCell = document.createElement("th");
		dataCell.classList.add("team2header","th");
		dataCell.innerHTML = "Team 2";
		header .appendChild(dataCell);
		return header;
	},

	getDataRow: function(date, league, leagueImageURL, team1,team1Url, team2, team2Url){
		let row = document.createElement("tr");

		row.appendChild(this.getDateCell(date));

		if(this.config.leagueAsImage){
			row.appendChild(this.getLeagueAsImage(leagueImageURL));
		}else{
			row.appendChild(this.getLeagueAsText(league));
		}

		if(this.config.teamAsImage){
			row.appendChild(this.getTeamAsImage(team1Url));
		}else{
			row.appendChild(this.getTeamAsText(team1));
		}
		row.appendChild(this.getVersusCell());
		if(this.config.teamAsImage){
			row.appendChild(this.getTeamAsImage(team2Url));
		}else{
			row.appendChild(this.getTeamAsText(team2));
		}

		return row;
	},

	getDateCell: function(date){
		let dataCell = document.createElement("td")
		dataCell.classList.add("dateCell","td");
		dataCell.innerHTML = date;
		return dataCell;
	},

	getVersusCell: function(){
		let versusCell = document.createElement("td");
		versusCell.classList.add("versuscell", "td");
		versusCell.innerHTML = "VS"
		return versusCell;
	},

	getLeagueAsImage: function(iconUrl){
		let dataCell = document.createElement("td");
		dataCell.classList.add("leagueCell", "td");
		let image = document.createElement("img");
		image.classList.add("leagueIcon", "img");
		image.setAttribute("src", iconUrl);
		dataCell.appendChild(image);
		return dataCell;
	},

	getLeagueAsText: function(leagueText){
		let dataCell = document.createElement("td");
		dataCell.classList.add("leagueCell","td");
		dataCell.innerHTML = leagueText;
		return dataCell;

	},

	getTeamAsImage: function(iconUrl){
		let dataCell = document.createElement("td");
		dataCell.classList.add("teamIconCell", "td");
		let image = document.createElement("img");
		image.classList.add("teamIcon", "img");
		image.setAttribute("src", iconUrl);
		dataCell.appendChild(image);
		return dataCell;
	},

	getTeamAsText: function(teamText){
		dataCell = document.createElement("td");
		dataCell.classList.add("teamText");
		dataCell.innerHTML = teamText;
		return dataCell;
	}
});
